working/0546-patterns/records-feature-specification.md

```diff
@@ -283,34 +323,88 @@ fields are) and collection literals.
+
 ## Runtime semantics
 
 ### Records
 
-#### Members
+#### Field getters
 
 Each field in the record's shape exposes a corresponding getter. Invoking that
 getter returns the value provided for that field when the record was created.
 Record fields are immutable and do not have setters.
 
-The `toString()` method's behavior is unspecified.
+#### `toString()`
+
+In debug builds, the `toString()` method converts each field to a string by
+calling `toString()` on its value and prepending it with the field name followed
+by `: ` if the field is named. It concatenates these with `, ` as a separator
+and returns the resulted surrounded by parentheses. For example:
+
+```dart
+print((1, 2, 3).toString()); // "(1, 2, 3)".
+print((a: 'str', 'ing').toString()); // "(a: str, int)".
+```
+
+The order that named fields appear and how they are interleaved with positional
+fields is unspecified. Positional fields must appear in position order. *This
+gives implementations freedom to choose a canonical order for named fields
+independent of the order that the record was created with.*
+
+In a release or optimized build, the behavior of `toString()` is unspecified.
+*This gives implementations freedom to discard the full names of named fields in
+order to reduce code size.* Users should only use `toString()` on records for
+debugging purposes. They are strongly discouraged from parsing the results of
+calling `toString()` or relying on it for end-user visible output.
 
 #### Equality
 
-Records behave similar to other primitive types in Dart with regards to
-equality. They implement `==` such that two records are equal iff they have the
-same shape and all corresponding pairs of fields are equal. Fields are compared
-for equality by calling `==` on the corresponding field values in the same
-order that `==` was called on the records.
+The `==` method on record `r` with right operand `o` is defined as:
+
+1.  If `o` is not a record with the same shape as `r` then return `false`.
+
+1.  For each pair of corresponding positional fields `rf` and `of` in position
+    order:
+
+    1.  If `rf == of` is `false` then return `false`.
+
+1.  For each pair of corresponding named fields `rf` and `of`, in unspecified
+    order:
+
+    1.  If `rf == of` is `false` then `false`.
+
+1.  Else, `true`.
+
+*The order that fields are iterated is potentially user-visible since
+user-defined `==` methods can have side effects. Most well-behaved `==`
+implementations are pure. The order that named fields are visited is
+deliberately left unspecified so that implementations are free to canonicalize
+their order.*
 
 ```dart
 var a = (x: 1, 2);
 var b = (2, x: 1);
 print(a == b); // true.
 ```
 
-The implementation of `hashCode` follows this. Two records that are equal must
-have the same hash code.
+The implementation of `hashCode` follows this. The hash code returned should
+depend on the field values such that two records that compare equal must have
+the same hash code.
```

Comment:
This implies that either (1) `hashCode` visits fields in some canonical order, or (2) the hash function is weakened so that it cannot distinguish permutations of named field values.

This is almost, but not quite, saying that the field order should be canonicalized. Perhaps we should go as far as to specify there is a canonicalization, and perhaps what it is.

Reply:
Not sure it implies that.
We do not allow equality between different record shapes, so as long as the order of fields used for the hashCode is consistent within the same shape, the order itself doesn't matter.
If two records are equal, the using `Object.hash(...)` on the field values in *any* order works, as long as it's the same order for all objects that has that shape. (This assumes that fields being equal implies their hash codes being equal to begin with.)

We do not promise that the `hashCode` is consistent between different executions of the same program, so precisely how we compute it shouldn't matter.

Reply:
Lasse has it right. It doesn't require hashing the fields in a canonical order across all record shapes, just that within a single shape, there is a consistent process for hashing them.